streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,691 @@
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(
+        NUM_BROKERS,
+        Utils.mkProperties(Collections.singletonMap("auto.create.topics.enable", "false"))
+    );
+
+    private static String applicationId;
+    private final static int NUM_TOPIC_PARTITIONS = 2;
+    private final static String CONSUMER_GROUP_ID = "readCommitted";
+    private final static String MULTI_PARTITION_INPUT_TOPIC = "multiPartitionInputTopic";
+    private final static String MULTI_PARTITION_OUTPUT_TOPIC = "multiPartitionOutputTopic";
+    private final String storeName = "store";
+
+    private AtomicBoolean errorInjected;
+    private AtomicInteger commitRequested;
+    private Throwable uncaughtException;
+
+    private int testNumber = 0;
+
+    @Before
+    public void createTopics() throws Exception {
+        applicationId = "appId-" + ++testNumber;
+        CLUSTER.deleteTopicsAndWait(
+            MULTI_PARTITION_INPUT_TOPIC,
+            MULTI_PARTITION_OUTPUT_TOPIC,
+            applicationId + "-" + storeName + "-changelog"
+        );
+
+        CLUSTER.createTopic(MULTI_PARTITION_INPUT_TOPIC, NUM_TOPIC_PARTITIONS, 1);
+        CLUSTER.createTopic(MULTI_PARTITION_OUTPUT_TOPIC, NUM_TOPIC_PARTITIONS, 1);
+    }
+
+    @Test
+    public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
+        // We use two KafkaStreams clients that we upgrade from eos-alpha to eos-beta. During the upgrade,
+        // we ensure that there are pending transaction and verify that data is processed correctly.
+        //
+        // We either close clients cleanly (`injectError = false`) or let them crash (`injectError = true`) during
+        // the upgrade. For both cases, EOS should not be violated.
+        //
+        // The commit interval is set to MAX_VALUE and the used `Processor` request commits manually so we have full
+        // control when a commit actually happens. We use an input topic with 2 partitions and each task will request
+        // a commit after processing 10 records.
+        //
+        // 1.  start both clients and wait until rebalance stabilizes
+        // 2.  write 10 records per input topic partition and verify that the result was committed
+        // 3.  write 5 records per input topic partition to get pending transactions (verified via "read_uncommitted" mode)
+        //      - both pending transactions are based on task producers
+        // 4.  stop/crash the first client, wait until rebalance stabilizes:
+        //      - stop case:
+        //        * verify that the stopped client did commit its pending transaction during shutdown
+        //        * the second client will still have an pending transaction
+        //      - crash case:
+        //        * the second client will have two pending transactions
+        // 5.  restart the first client with eos-beta enabled and wait until rebalance stabilizes
+        //       - crash only: the rebalance should result in a commit
```

Comment:
Not sure why it's the case? I think the previous pending txn should have aborted in step 4.

Reply:
Good catch. The commit happens for both cases, because when we migrate the task to the restarted client, the "old" client would commit the task. Will update the comment.